baybe/surrogates/base.py

```diff
@@ -51,6 +44,13 @@ class Surrogate(ABC, SerialMixin):
     """Class variable encoding whether or not the surrogate supports transfer
     learning."""
 
+    # NOTE: This attribute holds a subclass-specific object whose type and use
+    #   also depends on the particular surrogate at hand. Currently, the contained model
+    #   is assumed ephemeral and thus not considered for serialization and object
+    #   comparison via the equality operator.
+    _model: Any = field(init=False, default=None, eq=False)
```

Comment:
As you can see, I've pulled up `_model` to the base class, which can be considered a pure design choice and is debatable. 

**Pros:**
* We can impose some common settings (i.e. `eq` checks and serialization can now be disabled on the class level)
* It enables additional common logic such as the new "is fitted" check

**Cons:**
* The class actually does not use it
* It's not clear if all surrogates will ever actually use such an attribute. Having it here would simply assume they do. But after all, it's the responsibility of the classes to decide if they need it and what they do with it. For instance, the `MeanPredictionSurrogate` only needs to store a float and, who nows, other surrogates might not want to store anything or would prefer to use even additional fields. 
* We lose the specific type hints of the subclasses. There might be ways to override those in the subclass definitions, thought.
* RE "serialization" from "Pros": I guess in the long run we actually **do want** to serialize the trained model!?

Anyway, I think for now this solution provides a good middle ground and we can always adjust (it's a private field after all). But let me know what you think

Reply:
in a sense this is reminiscent of the issue with recommenders we had, ie some having internal models and some not. Would an intermediate abstract class derived from `Surrogate` (which then would require a model) be an option?

Reply:
also: not sure why these type hints are so important to you
Eg `Mother` class that has attribute `model: HypotheticalSKLearnModelBaseClass` and a `Child` class that uses `self.model = RandomForest()` is a valid combo which does not require setting the tye hint for the derived class anywhere, its simply not `necessary`. Given that we have additional options to provide model parameters to these by users, we should however mention the exact class used (and link) to the corresponding sklearn (or wherever the model used comes from) doc in the derived surrogates doc strings

Reply:
+1 for intermediate class

Reply:
I'm fine with rolling back, but I think the discussion went into a slightly wrong direction:
* Quick info for @AVHopp: while going back would mean that we again have exact type hints for the subclasses, please note that at least currently they are disabled for other reasons (limitation in `cattrs` that I identified, might file a PR for that). But that's just on the side...
* The case I mentioned above where we have a surrogate that doesn't have a model was rather hypothetical and I wanted to ask you if you can see such a case. In fact, I can't at the moment. And in fact, it would mean "a model (that's literally what a surrogate is) that doesn't rely on a model" ... rather controversial 😄 What we DO have currently is the `MeanPredictionModel` that only stores a float in that field – technically still a "model" though. So introducing an intermediate class seems like the absolute wrong direction to me, honestly (with zero use cases so far and probably none in the future)
* Much more likely could it be that models need to store more than one attribute to represent their model. In fact, this currently happens `autoscale`, where an additional scaler needs to be stored. But one could then argue just as well to simply use a tuple and assign that to the model field (though it requires modifying the type of the field in the decorator then)....

So ... no idea what's best. But strong vote against intermediate model :D

Reply:
If we cannot have type hints in general for this, I'd appreciate that this is mentioned and explained somewhere since I was not aware of this - for me, stuff like that is not "just on the side", it is important.

Regarding the importance of type hints (in general, ignoring the fact that we cannot use them here): My interpretation so far was that we do care quite a lot about type hints in general - we have explicit checks for them and we want to make sure that we use them properly, so using `Any` should only done when **absolutely necessary** (which seems to be the case here) - that is at least my opinion and what I thought we also agreed upon, but maybe that is yet another thing were I am wrong...

Regarding the other points: "Much more likely could it be that models need to store more than one attribute to represent their model" - but isn't that a point pro intermediate classes? What I read from your answer(s) is that there are already cases where claiming that something is "a model" is a bit of a stretch (for me, a float is not a model), and that we already have models that need to store more than just a model? How is having intermediate classes not the way to go here, in particular if we decided to do exactly that for `Recommenders` as pointed out by @Scienfitz ?

Reply:
ok i also misread your comment and thought your making the point that we `already` have a usecase for a modeless surrogate :D

it seems the surrogates are so flexible that it doesnt seem right to require them to have `model` but instead leave them free and implement stuff on their own.

In that scenario I think all of your pros are lost as its pointless to compare them or have `is_fitted` because what is fitted?

considering your own pro/cons list already are outweighed by cons why did you choose this design?

Reply:
Mostly because of minor things, e.g. the usage in the decorators. But let's discuss this in the meeting today 👍🏼

Reply:
OK, as discussed in the meeting, will remove `_model` again from base class for now and we'll tackle this stuff separately as preparation for the scaling